### ISSUE TITLE
Hitting space or arrows on iPad wiggles the page

### ISSUE THREAD

**Issue by wainwmr** [wainwmr/spem-player#10]
Need to swallow the bubbling of keyboard events that are handled by the canvas.

**Comment by wainwright1000** [comment]
## Description

On iPad (and potentially other touch devices with external keyboards), pressing Space or Arrow keys while interacting with the application causes the page to scroll or wiggle. The application handles these keys for playback and navigation controls, but the default browser behaviour is not fully suppressed, allowing the event to bubble and trigger page scrolling.

## Root Cause

The `keyboardTapped` handler in `index.ts` is attached to `document` and processes Space, Arrow keys, Enter, digit keys, and letter keys for application controls. While `e.preventDefault()` is called for Space, ArrowRight, and ArrowLeft (non-modifier), it is missing for ArrowUp, ArrowDown, Enter, and other handled keys. More importantly, `e.stopPropagation()` is never called, so the events continue to bubble up the DOM and may trigger default browser scrolling or focus behaviour.

## Recommended Fix

In `index.ts`, modify `keyboardTapped` to fully swallow all keyboard events that the application handles:

1. After determining that a key will be handled (and after the input-field / composing / control-widget guard clauses), call both `e.preventDefault()` and `e.stopPropagation()` before executing the key's action.
2. Ensure this applies to all handled keys: Space, Enter, ArrowUp, ArrowDown, ArrowLeft, ArrowRight, digit keys (1-8), and letter keys (S, A, T, R, B, V, M, D, X, / with Shift).
3. For modifier-key combinations (Ctrl/Meta + ArrowLeft/ArrowRight), also call `preventDefault()` and `stopPropagation()` so the page does not navigate.

A clean implementation is to add a single `e.preventDefault(); e.stopPropagation();` call at the top of the handler after the guard clauses, before any key-specific logic, since all keys that reach that point are explicitly handled.

## Test Plan

- Manual test on iPad with external keyboard: verify Space, all four Arrow keys, and Enter do not cause page scroll or wiggle.
- Desktop browser test: verify Arrow keys and Space still control playback and navigation correctly.
- Verify that typing in input fields (e.g., any future text inputs) is not affected.
- Verify that Ctrl/Meta + Arrow keys do not trigger browser back/forward navigation.

## Dependencies

- None; pure JavaScript/TypeScript change in `index.ts`.

## Notes

- The issue title mentions iPad specifically, but the same bubbling problem could affect any platform where keyboard events propagate to the document level.
- This is a follow-up to the existing keyboard handling code at `index.ts:210`.

**Comment by wainwright1000** [comment]
## Respecification

The previous specification lacked three required sections. The following adds **Current state**, **Scope**, and **Open questions**, and corrects the recommended fix.

### 1. Description

On iPad (and potentially other touch devices with external keyboards), pressing Space or Arrow keys while interacting with the application causes the page to scroll or wiggle. The application handles these keys for playback and navigation controls, but the default browser behaviour is not fully suppressed, allowing the event to bubble and trigger page scrolling.

### 2. Current state

- **Event listener attachment:** `keyboardTapped` is registered on `document` at `index.ts:429`:
  ```typescript
  document.addEventListener("keydown", keyboardTapped as (e: KeyboardEvent) => void);
  ```
- **Handler location:** `keyboardTapped` function spans `index.ts:210-318`.
- **Guard clauses (lines 211-229):**
  - Returns early if `e.target` is null or not an Element.
  - Returns early if the target has class `control` (prevents interference with UI widgets).
  - Returns early if the user is composing text (`e.isComposing || e.keyCode === 229`).
- **Modifier branch (lines 230-244):** Handles `Ctrl`/`Meta` + `ArrowLeft`/`ArrowRight` for seek navigation. Does **not** call `preventDefault()` or `stopPropagation()`.
- **Enter branch (lines 246-248):** Toggles play/pause. Does **not** call `preventDefault()` or `stopPropagation()`.
- **Space branch (lines 250-258):** Checks if target is an input/textarea/select and returns early if so. Toggles play/pause. Calls `e.preventDefault()` at line 257 but **not** `stopPropagation()`.
- **Switch statement (lines 260-316):** Handles digit keys (1-8), letter keys (S, A, T, R, B, V, M, D, X), Slash, Escape, and Arrow keys.
  - `ArrowRight` (line 295): calls `e.preventDefault()` at line 298.
  - `ArrowLeft` (line 300): calls `e.preventDefault()` at line 303.
  - `ArrowUp` (line 308): does **not** call `preventDefault()`.
  - `ArrowDown` (line 305): does **not** call `preventDefault()`.
  - All other cases in the switch: do **not** call `preventDefault()` or `stopPropagation()`.
- **`stopPropagation()` is never called** anywhere in `keyboardTapped`.

### 3. Root cause

While `e.preventDefault()` is called for `Space`, `ArrowRight`, and `ArrowLeft`, it is missing for `ArrowUp`, `ArrowDown`, `Enter`, and all letter/digit keys. More importantly, `e.stopPropagation()` is never called, so the events continue to bubble up the DOM. On iPad with an external keyboard, bubbling keyboard events can trigger default browser scrolling or focus behaviour, causing the page to wiggle.

### 4. Scope

**In scope:**
- Add `e.preventDefault()` and `e.stopPropagation()` to all handled keyboard events in `keyboardTapped` (`index.ts:210-318`).
- Ensure input fields, textareas, selects, and composition mode remain unaffected.

**Out of scope:**
- Touch scroll bounce (separate issue, recently fixed in commit `577016e`).
- Refactoring the keyboard handler into a separate module.
- Changing which keys are handled or adding new keyboard shortcuts.

### 5. Recommended fix

Restructure `keyboardTapped` so that every explicitly handled key calls both `preventDefault()` and `stopPropagation()`, while unhandled keys pass through normally.

**Option A: add to each branch (safer)**

Add `e.preventDefault(); e.stopPropagation();` to:
- The `metaKey`/`ctrlKey` branch after the switch, before the `return` at line 244.
- The `Enter` branch at line 247, before the `return` at line 248.
- The `Space` branch at line 256, alongside the existing `preventDefault()`.
- Each case in the main `switch` statement (lines 260-316), before the `break`.
- The `default` case must remain untouched so unmapped keys do not swallow browser shortcuts.

**Option B: extract handler and use a Set (cleaner)**

Replace the switch statement with a Set of handled key codes. After the guard clauses, check if `e.code` is in the Set; if so, call `preventDefault()` and `stopPropagation()`, then dispatch to the action. This avoids sprinkling the calls across 15+ case labels.

Example sketch:

```typescript
const handledKeys = new Set([
  'Enter', 'Space', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight',
  'Digit1', 'Digit2', 'Digit3', 'Digit4', 'Digit5', 'Digit6', 'Digit7', 'Digit8',
  'KeyS', 'KeyA', 'KeyT', 'KeyR', 'KeyB', 'KeyV', 'KeyM', 'KeyD', 'KeyX',
  'Slash', 'Escape',
]);

if (handledKeys.has(e.code)) {
  e.preventDefault();
  e.stopPropagation();
}
```

Then execute the existing switch logic. The modifier branch (`Ctrl`/`Meta` + Arrow) should also call both methods since it is explicitly handled.

### 6. Test plan

1. **Unit test:** Extend `src/test/keyboard.test.ts`. Add assertions that `preventDefault()` and `stopPropagation()` are called for handled keys. Spy on the event methods:
   ```typescript
   const event = new KeyboardEvent('keydown', { code: 'Space', bubbles: true });
   const preventDefaultSpy = vi.spyOn(event, 'preventDefault');
   const stopPropagationSpy = vi.spyOn(event, 'stopPropagation');
   document.dispatchEvent(event);
   expect(preventDefaultSpy).toHaveBeenCalled();
   expect(stopPropagationSpy).toHaveBeenCalled();
   ```
2. **Manual iPad test:** Connect an external keyboard to an iPad. Press Space, ArrowUp, ArrowDown, ArrowLeft, ArrowRight, Enter, and digit keys. Verify the page does not scroll or wiggle.
3. **Regression test:** Verify that typing in `input`, `textarea`, and `select` elements still works correctly (the guard clauses at lines 223-229 must remain intact).
4. **Regression test:** Verify that unhandled keys (for example unmapped letter keys) still trigger their default browser behaviour.

### 7. Dependencies

- None.

### 8. Open questions

1. Option A (add to each branch) is minimal and low-risk. Option B (Set + centralised swallow) is cleaner but touches more lines. Which approach does Mark prefer?
2. Should the `console.log` at line 315 (default case) be removed as part of this cleanup, or left for debugging unknown keys?

**Comment by wainwright1000** [comment]
Checked current state of `keyboardTapped` in `index.ts` (HEAD 3007812, lines 229–346).

**preventDefault():**
- Present: Space, ArrowRight, ArrowLeft, KeyF
- Missing: Enter, ArrowUp, ArrowDown, digit keys (1–8), letter keys (S, A, T, R, B, V, M, D, X), Slash, Escape, Ctrl/Meta + Arrow combinations

**stopPropagation():**
- Missing for all handled keys — never called anywhere in `keyboardTapped`.

The partial preventDefault coverage that exists does not suppress bubbling, so iPad page wiggle remains possible for any key that lacks both calls. This ticket is not yet resolved.

- Kimi

**Comment by wainwright1000** [comment]
# VERA-10 Original Report (cycle 1)

Mode: work (Vera ran during initial development, before PR open)
Cycle: 1
Generated: 2026-05-28 00:05

See also: [Final Synthesis (cycle 1)](https://github.com/wainwmr/spem-player/issues/10#issuecomment-4561299282)

## Raw agent reports (pass 1)

### code-reviewer

**Critical:**

- `index.ts:283-288` — Browser shortcuts Cmd+S/Cmd+F/Cmd+A/etc. now silently swallowed when document body is focused. `HANDLED_KEYS` contains `KeyS`, `KeyA`, `KeyD`, `KeyF`, `KeyR`, `KeyT`, `KeyB`, `KeyV`, `KeyM`, `KeyX`, `Digit1`...`Digit8`, `Enter`, `Slash`. When the user presses Cmd+S (save page), Cmd+F (find), etc. without focus inside an input, `HANDLED_KEYS.has(e.code)` is true and `preventDefault()` is called. The Ctrl/Cmd branch has no case for these codes — the shortcut is silently eaten. Fix: gate on `!e.metaKey && !e.ctrlKey`, or narrow set to keys that actually cause iPad scrolling.
- `index.ts:283-288` — Enter on focused buttons (e.g. feedback `#feedback-submit`, `#feedback-cancel`) is suppressed. `HANDLED_KEYS` includes `Enter`. The `isInputLike` guard at index.ts:246-247 exempts inputs/textareas/select/elements with class `control` but NOT native `<button>`. Per HTML spec, `preventDefault` on Enter keydown suppresses the synthetic click. Tab + Enter on feedback Send button: starts/stops playback instead of submitting. Keyboard accessibility regression.

**Important:**

- `index.ts:283-288` — Escape inside an input now calls `preventDefault()`, losing native browser behaviour (cancel IME composition; macOS field revert). Drop `Escape` from `HANDLED_KEYS`.
- `index.ts:283-288` — `e.stopPropagation()` is broader than needed. Handler is bubble-phase on document; stopPropagation only blocks window-level listeners (none exist). Remove.
- `src/test/keyboard.test.ts:173-339` — Tests assert spy calls but not the actual behaviour, nor the no-regression cases. None of the new tests would catch the broken Cmd+S, Cmd+F, or Enter-on-button.

### pr-test-analyzer

**No critical blocker** but 9 important findings:

- [A] sev 7 — `keyboard.test.ts:323` input-focused test only covers `<input>` via JSDOM bubble; doesn't lock in the `<div class="control">` branch.
- [B] sev 7 — missing Escape-in-input test. Handler has `if (e.code !== "Escape") return;` at index.ts:264 — Escape IS handled even in inputs (to close modals). Untested.
- [C] sev 6 — modifier-key matrix thin. Only Ctrl+ArrowRight tested. Missing Ctrl+Space, Shift+Arrow, Alt+Arrow.
- [D] sev 6 — IME composition path untested (`isComposing: true`).
- [E] sev 5 — key repeat (`repeat: true`) untested.
- [F] sev 5 — `e.target === null` and non-Element target untested.
- [G] sev 6 — 11 of 13 new tests are near-identical copy-paste using `spyOn(event, "preventDefault")` — implementation, not behaviour. Should use `event.defaultPrevented`. Could collapse to `it.each` table.
- [H] sev 5 — assertions don't catch a wiggle regression (preventDefault is necessary but not sufficient).
- [I] sev 4 — negative test uses `KeyZ`; `Tab` would be more interesting.

### silent-failure-hunter

**Critical:**

- `Slash` without Shift is preventDefault'd but performs no app action — silent swallow. `HANDLED_KEYS` includes `"Slash"`, so `/` is preventDefault'd. But the switch case only fires `showHelp()` when `e.shiftKey` (i.e. for `?`). Plain `/` outside an input vanishes silently (breaks Firefox quick-find).
- `Enter` swallowed globally — breaks buttons, links, form submission outside `isInputLike` filter. (Concurs with code-reviewer.)

**Important:**

- The "input focused" test (keyboard.test.ts:325-339) is a false-positive in JSDOM — only checks Space on `<input>`. Doesn't verify Enter on `<button>`, `<a>`, `[role=button]`, `[tabindex="0"]`.
- No `e.repeat` guard on Ctrl+ArrowLeft/Right. Holding fires rapid `setBar` calls.
- Listener leak risk on test re-imports (`vi.resetModules()` re-registers without removeEventListener).
- Redundant `isModifierSeek` check (cosmetic).

### type-design-analyzer

**PASS** from a type-design perspective. No critical/important. One nit:

- `HANDLED_KEYS` is inferred as `Set<string>`. Tighter declaration (`ReadonlySet<string>` with `as const`) would catch typos. Pre-existing duplication between the set and the switch.

### comment-analyzer

**Important:**

- `index.ts:229-254` — `HANDLED_KEYS` and the preventDefault block are the entire payload of the fix, yet there is no comment explaining the why (iPad page-wiggle; stopPropagation rationale; modifier-arrow special case; warning that new switch keys must also be added to the set). Single most important "why" in the patch and is undocumented. Future maintainer adding a new key to the switch will likely forget the set, silently regressing the iPad fix.
- `index.ts:226` — stale section header `// Keyboard events (wasd)` now misleading (covers Space, Enter, Arrows, Digits 1-8, S/A/T/R/B, V/M/D/F/X, Slash, Escape).
- `index.ts:246-247` — pre-existing Escape special case (`if (e.code !== "Escape") return`) is now load-bearing for the input-focused test. Worth a comment noting that Escape is intentionally allowed through.

**Suggestions:**

- `keyboard.test.ts:172` — `describe("Keyboard event swallowing (#10)")` will rot when issue closes. Use behavioural name; put `#10` in a one-line comment if needed.
- `keyboard.test.ts:172-339` — 13 tests with zero comments. Negative tests at lines 318 and 329 are the load-bearing ones — deserve one-line comment each explaining the invariant guarded.

- Claude (Vera gate runner)

**Comment by wainwright1000** [comment]
# VERA-10 Final Synthesis (cycle 1)

Mode: work (Vera ran during initial development, before PR open)
Cycle: 1
Generated: 2026-05-28 00:05
Last run:  2026-05-28 00:05

See also: [Original Report (cycle 1)](https://github.com/wainwmr/spem-player/issues/10#issuecomment-4561245376)

## Summary

One Vera cycle, two toolkit passes. Pass 1 produced 14 findings (3 critical, 11 important, 5 suggestions). All three criticals (Cmd/Ctrl+letter swallowed, Enter on focused button, Slash without Shift) were addressed in commit `f789a65` by narrowing `HANDLED_KEYS` → `SCROLL_KEYS = {Space, ArrowUp, ArrowDown, ArrowLeft, ArrowRight}` and gating on `!e.metaKey && !e.ctrlKey`. Most importants addressed in the same commit (test-block rewrite to `defaultPrevented`, drop `stopPropagation`, comment improvements). Pass 2 produced 4 new important: one pre-existing-and-out-of-scope (Enter side-effect → deferred to refactor doc), one test gap (Cmd+Arrow not tested), and two self-inflicted comment-drift items from my cycle-1 fix — same shape as #369 cycle 1. The three actionable findings landed in commit `36c0619`; pass 3 was skipped because the only code change was test duplication (no production logic).

Headline: the original commit's `HANDLED_KEYS` set was three times broader than the ticket required. Vera caught a genuine accessibility regression (Tab + Enter on feedback Submit) and a real browser-shortcut regression (Cmd+S, Cmd+F, Cmd+A swallowed) on what looked like a small iPad-scroll fix. Three of four behaviour-aware agents independently converged on the same root cause.

Six commits total on the branch: original (`64bb28d`), Vera ORIGINAL + synthesis stub (`880245b`, `91a6eaa`), cycle-1 fix (`f789a65`), synthesis update (`63d6fe4`), cycle-2 fix (`36c0619`). Plus close-out commit (deletes both VERA files).

## Findings

### 10-01 — critical: `HANDLED_KEYS` swallows Cmd/Ctrl+letter browser shortcuts

> **code-reviewer, index.ts:283-288:**
> `HANDLED_KEYS` contains `KeyS`, `KeyA`, `KeyD`, `KeyF`, `KeyR`, `KeyT`, `KeyB`, `KeyV`, `KeyM`, `KeyX`, `Digit1..8`, `Enter`, `Slash`. When the user presses Cmd+S (save page), Cmd+F (find), etc. without focus inside an input, `HANDLED_KEYS.has(e.code)` is true and `preventDefault()` fires. The Ctrl/Cmd branch has no case for these codes — the shortcut is silently eaten. Real user-observable regression.

**Bob's triage:** real defect, address now. Narrow `HANDLED_KEYS` to just the keys that cause iPad page-wiggle (`Space`, `ArrowUp`, `ArrowDown`, `ArrowLeft`, `ArrowRight`) and gate on `!e.metaKey && !e.ctrlKey` so browser shortcuts pass through. The over-broad set was the root cause.

**Resolution:** addressed (commit `f789a65`). `HANDLED_KEYS` replaced with `SCROLL_KEYS = {Space, ArrowUp, ArrowDown, ArrowLeft, ArrowRight}`, gated on `!e.metaKey && !e.ctrlKey`. Browser shortcuts (Cmd+S/F/A/R/P, Ctrl+S/F/A/R/P) now pass through cleanly.

### 10-02 — critical: `Enter` swallowed on focused `<button>`/`<a>`

> **code-reviewer + silent-failure-hunter, index.ts:283-288:**
> `HANDLED_KEYS` includes `Enter`. The `isInputLike` guard exempts inputs/textareas/select/`.control` but NOT native `<button>`, `<a>`, `[role=button]`. Tab + Enter on the feedback Send/Cancel button: starts/stops playback instead of submitting. `stopPropagation` also kills component-level Enter handlers (silently muted). Keyboard accessibility regression.

**Bob's triage:** real defect, address now. Subsumed in the 10-01 narrowing — `Enter` drops out of `HANDLED_KEYS` entirely. Buttons regain their native click-on-Enter behaviour.

**Resolution:** addressed (commit `f789a65`). `HANDLED_KEYS` replaced with `SCROLL_KEYS = {Space, ArrowUp, ArrowDown, ArrowLeft, ArrowRight}`, gated on `!e.metaKey && !e.ctrlKey`. Browser shortcuts (Cmd+S/F/A/R/P, Ctrl+S/F/A/R/P) now pass through cleanly.

### 10-03 — critical: `Slash` without Shift silently swallowed

> **silent-failure-hunter, index.ts:283-288 vs index.ts:308-312:**
> `HANDLED_KEYS` contains `"Slash"`, so `/` is preventDefault'd. But the switch only fires `showHelp()` when `e.shiftKey`. Plain `/` outside an input vanishes silently. Breaks Firefox quick-find.

**Bob's triage:** real defect, address now. Subsumed in 10-01 narrowing — `Slash` drops out.

**Resolution:** addressed (commit `f789a65`). `HANDLED_KEYS` replaced with `SCROLL_KEYS = {Space, ArrowUp, ArrowDown, ArrowLeft, ArrowRight}`, gated on `!e.metaKey && !e.ctrlKey`. Browser shortcuts (Cmd+S/F/A/R/P, Ctrl+S/F/A/R/P) now pass through cleanly.

### 10-04 — important: Escape inside input now preventDefault'd

> **code-reviewer, index.ts:283-288:**
> The `isInputLike` guard deliberately lets Escape through to close modals. The new swallowing block then calls `preventDefault()` on Escape, losing native browser behaviour (cancel IME composition; macOS field revert).

**Bob's triage:** subsumed in 10-01 narrowing — `Escape` drops out of `HANDLED_KEYS`. The modal-close behaviour stays in the existing Escape switch case.

**Resolution:** [auto-addressed by 10-01 fix]

### 10-05 — important: No `e.repeat` guard on Ctrl+Arrow seek

> **silent-failure-hunter, index.ts:283-289:**
> No `e.repeat` check. Holding the arrow key fires repeated `controls.pause()` + `setBar(...)` calls at the OS repeat rate.

**Bob's triage:** defensive nit (the rapid seek is arguably intended scrubbing behaviour, not a defect). Defer.

**Resolution:** deferred — refactor item; current behaviour may be intentional; user may revisit.

### 10-06 — important: Input-focused test is false-positive in JSDOM

> **silent-failure-hunter, keyboard.test.ts:325-339:**
> Only covers Space on `<input>`. Doesn't verify Enter on `<button>`, `<a>`, `[role=button]`, `[tabindex="0"]`.

**Bob's triage:** real defect (test gap), address now. After 10-01 narrowing, Enter is no longer swallowed at all, so the button case becomes irrelevant. But the test should still assert defaultPrevented stays false for the new wider non-handled key set.

**Resolution:** addressed (commit `f789a65`). Test block rewritten to use `event.defaultPrevented` with table-driven `it.each` cases; negative cases added for Cmd+letter, Enter on `<button>`, Slash, Escape, Tab, F5, and IME composition.

### 10-07 — important: `stopPropagation()` unnecessary at document-level bubble phase

> **code-reviewer, index.ts:287:**
> Handler registered with `document.addEventListener("keydown", ...)` (bubble phase). By the time document sees the event it has already bubbled past `document.body`, so `stopPropagation` only blocks window-level listeners (none exist). Misleading.

**Bob's triage:** real defect (misleading code), address now. Drop the `stopPropagation()` call.

**Resolution:** addressed (commit `f789a65`).

### 10-08 — important: Tests assert spy calls, not behaviour

> **code-reviewer + pr-test-analyzer, keyboard.test.ts:173-339:**
> 11 of 13 new tests are near-identical copy-paste using `spyOn(event, "preventDefault")`. Couples to implementation. Use `event.defaultPrevented` for behavioural assertion.

**Bob's triage:** real defect, address now. Rewrite the describe block to use `defaultPrevented` and collapse to a table-driven `it.each` for the positive cases, plus distinct `it` blocks for the load-bearing negative cases (Cmd+S, Cmd+F, Enter-on-button, Slash-without-Shift).

**Resolution:** addressed (commit `f789a65`).

### 10-09 — important: Escape-in-input case missing tests

> **pr-test-analyzer, keyboard.test.ts:**
> Handler has `if (e.code !== "Escape") return;` so Escape IS handled even when an input is focused. Untested.

**Bob's triage:** real defect (test gap), address now. Add explicit test that Escape inside an `<input>` does NOT early-return (the in-input guard is bypassed for Escape so modal-close handlers run).

**Resolution:** addressed (commit `f789a65`).

### 10-10 — important: Modifier matrix thin

> **pr-test-analyzer:**
> Only Ctrl+ArrowRight tested. Missing Ctrl+Space (now relevant: SHOULD NOT preventDefault under narrowed scope), Shift/Alt+Arrow.

**Bob's triage:** real defect (test gap), address now. Add: Cmd+S not preventDefault'd, Cmd+F not preventDefault'd, Ctrl+ArrowRight IS preventDefault'd (isModifierSeek special case), Shift+ArrowRight IS preventDefault'd.

**Resolution:** addressed (commit `f789a65`).

### 10-11 — important: IME composition path untested

> **pr-test-analyzer, index.ts:281:**
> Handler returns early on `e.isComposing || e.keyCode === 229`. If a future refactor moves the swallow block before this guard, Chinese/Japanese input would break silently.

**Bob's triage:** defensive nit. Address now (one test, cheap).

**Resolution:** addressed (commit `f789a65`). Single IME test added pinning the early-return order.

### 10-12 — important: `HANDLED_KEYS` missing "why" comment

> **comment-analyzer, index.ts:229-254:**
> The single most important "why" in the patch is undocumented. Future maintainer adding a new key to the switch will likely forget the set, silently regressing the iPad fix.

**Bob's triage:** real defect (maintainability), address now. Add a block comment naming the bug, the modifier-gate rationale, and the "add new keys here" warning.

**Resolution:** addressed (commit `f789a65`).

### 10-13 — important: Stale section header `// Keyboard events (wasd)`

> **comment-analyzer, index.ts:226:**
> Pre-existing header references "wasd" but the handler now covers a much wider key set. Misleading.

**Bob's triage:** real defect (comment drift, in-scope to fix here while editing nearby), address now. One-line update.

**Resolution:** addressed (commit `f789a65`).

### 10-14 — important: Escape exception in `isInputLike` load-bearing-but-undocumented

> **comment-analyzer, index.ts:246-247:**
> Pre-existing `if (e.code !== "Escape") return;` is now load-bearing for modal-close behaviour. Worth a comment.

**Bob's triage:** real defect (maintainability), address now. One-line comment.

**Resolution:** addressed (commit `f789a65`).

## Pass 2 findings

Pass 2 re-ran all five agents against the updated branch tip after the pass-1 fixes landed. Three agents returned clean (no new critical/important):

- **code-reviewer**: clean at confidence ≥ 80. Notes the Enter-on-button play/pause-toggle as a sub-threshold pre-existing observation (confidence ~40).
- **type-design-analyzer**: clean. Ratings improved across all four dimensions (Encapsulation 9, Invariant Expression 8, Usefulness 9, Enforcement 8).
- **silent-failure-hunter**: ONE new important → 10-20 below.
- **pr-test-analyzer**: ONE new important → 10-21 below.
- **comment-analyzer**: TWO new important → 10-22, 10-23 below.

### 10-20 — important: Enter on focused button still triggers play/pause as side effect

> **silent-failure-hunter (pass 2), index.ts:268-271 + isInputLike guard:**
> Pass 1's 10-02 was Enter being SWALLOWED on focused buttons. My fix removed Enter from the swallow set so `preventDefault` is no longer called. But the `if (e.code == "Enter")` block at line 268 still fires `controls.isPlaying() ? pause() : play()` whenever Enter is pressed and the early-return guard doesn't catch it. Pressing Tab + Enter on the feedback Submit button activates the button AND toggles playback as a side effect.

**Bob's triage:** real defect, but **pre-existing**. The play/pause-on-Enter behaviour was present before #10; my fix only unmasked it by no longer suppressing the button click. code-reviewer concurs (rated sub-threshold; "pre-existing behavior unchanged by this branch"). Out of #10 scope.

**Resolution:** deferred — file as refactor item in `wiki/refactor-index.ts.md` (or workbench item if a refactor report doesn't exist): "Enter on focused interactive elements (`<button>`, `<a>`) triggers global play/pause as a side effect because the `if (e.code == "Enter")` block runs before any button-specific guard. Fix shape: extend `isInputLike` to include `HTMLButtonElement` and `HTMLAnchorElement`, or gate the Enter block on `document.activeElement === document.body`."

### 10-21 — important: Cmd+Arrow seek untested (only Ctrl+Arrow tested)

> **pr-test-analyzer (pass 2), keyboard.test.ts:208-216:**
> The `isModifierSeek` branch is `(e.metaKey || e.ctrlKey) && ...`. Tests only exercise `ctrlKey: true`. macOS/iPad users press Cmd+Arrow. A future refactor that narrows `isModifierSeek` to ctrlKey only would silently break the iPad/macOS path and tests would stay green.

**Bob's triage:** real defect (test gap), address now. The iPad path is the original ticket motivation; Cmd+Arrow is the modifier users actually press on macOS.

**Resolution:** addressed (commit `36c0619`). Added an `it.each` cohort for `metaKey: true` alongside the existing `ctrlKey: true` block.

### 10-22 — important: Comments say "Ctrl+Arrow" but code allows Cmd+Arrow

> **comment-analyzer (pass 2), index.ts:233 + index.ts:273:**
> Both new comments call Ctrl+ArrowLeft/Right "the one modifier combination" the app owns. Code at line 277 (`e.metaKey || e.ctrlKey`) and the switch at line 291 accept both. Comments undercount.

**Bob's triage:** real defect (factual drift, self-inflicted by my cycle-1 fix), address now. One-line edit in two places.

**Resolution:** addressed (commit `36c0619`). Changed "Ctrl+ArrowLeft/Right" → "Cmd/Ctrl+ArrowLeft/Right" in both comment blocks.

### 10-23 — important: Test comment over-claims Escape case behaviour

> **comment-analyzer (pass 2), keyboard.test.ts:258-260:**
> Comment reads "The Escape switch case can still call preventDefault internally when it actually closes a modal" — the actual Escape case at `index.ts:343-346` does NOT call preventDefault.

**Bob's triage:** real defect (factual drift, self-inflicted), address now.

**Resolution:** addressed (commit `36c0619`). Comment rewritten to describe the real behaviour: "The top-level swallow block must not call preventDefault for it; the Escape switch case dismisses modals without preventDefault."

## Pass 3 — skipped

Per gate spec: documentation-and-test-only changes since the last clean pass (the only code change in `36c0619` was duplicating one `it.each` block to cover metaKey; no production logic changed). Pass 3 would only re-verify documentation accuracy on changes that comment-analyzer pass 2 just validated. Skip is appropriate.

## Suggestions (non-blocking)

### 10-15 — suggestion: `HANDLED_KEYS` could be `ReadonlySet<KeyboardEvent["code"]>` for tighter typing

> **type-design-analyzer, index.ts:229:**

**Resolution:** deferred — minor polish; tracked in post-merge refactor doc.

### 10-16 — suggestion: pre-existing dead branch `if (e === undefined || e.target === null)`

> **type-design-analyzer, index.ts:230-232:**

**Resolution:** deferred — pre-existing, out of scope.

### 10-17 — suggestion: `describe("Keyboard event swallowing (#10)")` will rot

> **comment-analyzer, keyboard.test.ts:172:**

**Resolution:** addressed in 10-01 fix — rewrite the describe with a behavioural title.

### 10-18 — suggestion: listener leak risk on test re-imports

> **silent-failure-hunter, index.ts:555-558:**

**Resolution:** deferred — tooling fix, broader than this ticket.

### 10-19 — suggestion: redundant `isModifierSeek` check

> **silent-failure-hunter:**

**Resolution:** kept — under the narrowed scope, `isModifierSeek` is the only branch that handles Ctrl+ArrowLeft/Right; the check is now load-bearing rather than redundant.

## Cross-cutting note

Five Vera agents, three flagged the same root cause from different angles: code-reviewer, silent-failure-hunter, and pr-test-analyzer all identified that `HANDLED_KEYS` is too broad. The fix scope is the user's original commit; the user explicitly chose to include the full handled-keys set. Bob's call to narrow stands on the strict reading of the ticket (Space + arrows cause iPad scroll) and the project's bug definition (user-observable failure during normal use).

- Claude (Vera gate runner)

**Comment by wainwright1000** [comment]
# VERA-10 Original Report (cycle 2)

Mode: work (Vera ran during initial development, before PR open)
Cycle: 2
Generated: 2026-05-28 13:35

See also: [Final Synthesis (cycle 2)](https://github.com/wainwmr/spem-player/issues/10#issuecomment-4564118201)
(Earlier cycles: [Synthesis (cycle 1)](https://github.com/wainwmr/spem-player/issues/10#issuecomment-4561299282))

## Raw agent reports (pass 1)

### code-reviewer

> ## Review summary
>
> Files reviewed:
> - `c:/Users/Andrew/Spem/index.ts` (lines 225-374 — keyboard event handler + new SCROLL_KEYS set)
> - `c:/Users/Andrew/Spem/src/test/keyboard.test.ts` (lines 173-312 — new "Keyboard event swallowing for iPad page-wiggle" suite)
>
> Verification run: 30 tests pass; `npx tsc --noEmit` and `npx eslint` clean on both files.
>
> This is a tight, narrow fix. The semantic shift from "swallow everything in the handler" to "swallow only scroll-causing keys" is correctly implemented and well-tested (positive and negative cases, modifier combinations, IME, input-focused).
>
> The PR description mentions `PageUp/PageDown/Home/End` in the swallow set, but the code includes only `Space` and the four arrow keys. The code's narrower set is the right call; consider tightening the PR description so reviewers don't go looking for the extra keys.
>
> ### Critical: None.
>
> ### Important
>
> **1. Misleading test comment — "previously silently swallowed" is factually wrong** (confidence 82)
> - File: `c:/Users/Andrew/Spem/src/test/keyboard.test.ts:259-260`
> - The comment reads: *"Plain `/` (no Shift) is not a help-modal trigger; previously it was silently swallowed which broke Firefox quick-find."* This is incorrect — on `origin/main`, the `case "Slash"` block only acts when `e.shiftKey` is true and never called `preventDefault()` for plain `/`. The pre-fix code did NOT swallow plain Slash.
> - Fix: rewrite the comment to describe the actual invariant being pinned. The test itself is still valuable (it guards against a future widening of `SCROLL_KEYS`); only the historical claim is wrong.
>
> ### Suggestions
>
> **2. Prettier formatting nit on `isPlainScrollKey`** (confidence 70)
> - File: `c:/Users/Andrew/Spem/index.ts:279-280`
> - `prettier --check` collapses these two lines into one.
>
> **3. PR description says `PageUp/PageDown/Home/End`; the code does not include them** (confidence 75 — process / description issue)
> - File: `c:/Users/Andrew/Spem/index.ts:236-242`
> - Recommend: amend the PR description.

### pr-test-analyzer

> ## Summary
>
> The test file is meaningfully scoped: it asserts the user-visible contract (`defaultPrevented`) rather than spying on `preventDefault`, parametrises positive and negative cases, and pins a non-obvious ordering invariant (IME composition before swallow).
>
> However, there is a **material contradiction between the ticket description and the implementation** that the tests do not catch.
>
> ### Critical
>
> **[C1] Severity 9 — Ticket claims `PageUp/PageDown/Home/End` are swallowed; the code does not, and no test asserts either way.**
>
> - Production code at `c:/Users/Andrew/Spem/index.ts:236-242`:
>   ```
>   const SCROLL_KEYS = new Set([
>     "Space", "ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight",
>   ]);
>   ```
>   `PageUp`, `PageDown`, `Home`, and `End` are absent.
> - The parametrised positive table at `keyboard.test.ts:192-205` mirrors the implementation rather than the ticket. If the ticket description is authoritative, the bug is unfixed.
>
> **Recommended action:** Either (a) add `PageUp`, `PageDown`, `Home`, `End` to `SCROLL_KEYS` and extend the test table, or (b) update the ticket and the PR description if these were deliberately deferred.
>
> ### Important
>
> **[I1] Severity 7 — No test covers Space *with* a modifier (Cmd+Space / Ctrl+Space).** Cmd+Space behaviour (Spotlight on macOS) must not be swallowed and must not toggle play/pause. The desired behaviour is entirely untested.
>
> **[I2] Severity 7 — Tests assert `defaultPrevented` but never assert that the action *also* fired (or didn't).** If a future refactor removed the `if (e.code == "Space")` branch entirely, the swallow would still fire and the test would still pass — but play/pause would be broken.
>
> **[I3] Severity 6 — `KeyF` lost its `preventDefault` — no test that plain `f` (no modifier) actually opens the feedback modal *and* doesn't trigger any browser default.**
>
> ### Suggestions
>
> **[S1] Inconsistent dispatch target in negative tests.**
>
> **[S2] The `control`-class guard at `index.ts:263` is not exercised.**
>
> **[S3] Escape-on-input is documented but not tested.** The actual new code path — Escape *on a focused textarea* — is untested.
>
> **[S4] `e.keyCode === 229` IME branch only partially covered.**
>
> **[S5] `defaultPrevented` for keys-that-are-handled-but-not-in-SCROLL_KEYS could be pinned.** (KeyD, KeyM, etc.)
>
> **[S6] Early returns at `index.ts:245-252` are dead-code-ish but untested.**
>
> ### Recommendation for the Vera gate
>
> **Do not pass cycle 2 until [C1] is resolved** — either the missing keys are added, or the ticket scope is explicitly narrowed in the PR description.

### silent-failure-hunter

> ## CRITICAL
>
> ### C1. `SCROLL_KEYS` set does not match the commit message / ticket scope — silent omission
>
> **Location:** `c:/Users/Andrew/Spem/index.ts:236-242` and commit `d169f16`
>
> The commit message states the swallow set is "scroll-causing keys only (Space, ArrowUp, ArrowDown, **PageUp, PageDown, Home, End**)". The actual `SCROLL_KEYS` set is Space + 4 arrows.
>
> This is the textbook silent-failure pattern: the contract written in the commit message is *wider* than the code's behaviour, with no test, log, or comment acknowledging the gap.
>
> **Recommendation:** either (a) add the four missing keys to `SCROLL_KEYS` *and* add `it.each` rows for them, or (b) rewrite the commit/PR description to truthfully name "Space and Arrow keys only", and explicitly document why PageUp/Home/End were ruled out.
>
> ## IMPORTANT
>
> ### I1. Silent early-return on non-Element target loses scroll-prevention
>
> **Location:** `c:/Users/Andrew/Spem/index.ts:249-252`. If a keydown fires with `e.target` being `document` or `window` (which can happen in iOS Safari with focus on `<body>`), the handler returns before swallowing the scroll key — re-introducing the wiggle. Worth verifying on a real iPad or adding a test that dispatches keydown directly to `document`.
>
> ### I2. `isInputLike` guard at line 263-265 silently swallows scroll-key preventDefault for input-focused users
>
> Intentional, pinned by test. But the block comment doesn't mention that scroll-key preventDefault is also bypassed for input-like targets — a future reader chasing I1 will silently break textarea typing. One-sentence comment addition recommended.
>
> ### I3. Test helper `dispatchKeydown` silently ignores `cancelable` override
>
> The spread-after-defaults pattern silently accepts a `cancelable: false` override. Low-grade footgun, not a defect.
>
> ## SUGGESTION
>
> ### S1. `isModifierSeek` accepts Alt/Shift+Arrow as a "plain scroll key" — not what the comment promises
>
> `isPlainScrollKey` requires only `!metaKey && !ctrlKey` — it permits `altKey` or `shiftKey`. The block comment says "only when no modifier is held", which is broader than the code.
>
> ### S2. `KeyboardEvent.code` is the only branch key — non-physical-keyboard input is silently uncategorised
>
> On iPad with software keyboards, autocorrect-injected events, and accessibility switch input can dispatch keydown events with `e.code === ""` and only `e.key` populated. Such events are silently let through.
>
> ### S3. No coverage of `Shift+Space` (page-up scroll on desktop browsers)
>
> ## SUMMARY
>
> One **critical** issue: `SCROLL_KEYS` is missing `PageUp`/`PageDown`/`Home`/`End` per the commit message. Two **important** issues around order-sensitivity of guards. Three **suggestions** about comment/code drift, `e.code` vs `e.key`, and missing test rows.

### type-design-analyzer

> ## Type: `SCROLL_KEYS` (file: `c:/Users/Andrew/Spem/index.ts:236-242`)
>
> ### Invariants identified
>
> - [A] Membership: each element should be a valid `KeyboardEvent.code` value.
> - [B] Scope: every member is a key whose browser default scrolls the page.
> - [C] Read-only: the set is constructed at module load and never mutated.
> - [D] Single source of truth: any key that needs `preventDefault` for scroll-suppression should appear here exactly once.
> - [E] Disjoint-with-shortcut policy: the set is consulted only on the `!metaKey && !ctrlKey` path.
>
> ### Ratings
>
> - **Encapsulation: 7/10** — `Set<string>` exposes the full mutating API. `ReadonlySet<string>` annotation would close the gap.
> - **Invariant Expression: 5/10** — the *type* says `Set<string>`; any string is admitted. There is no compile-time link between the set's members and the `e.code` literals used in the switch statements. A typo (`"Spcae"`) would compile and silently fail.
> - **Invariant Usefulness: 6/10** — the set captures the bug fix's intent. **But** the set is incomplete relative to the ticket description (PageUp/PageDown/Home/End absent).
> - **Invariant Enforcement: 6/10** — no enforcement that the set's members match the `case` labels in the switch below, nor that the codes are real DOM `code` values.
>
> ### Concerns
>
> **Important — ticket scope mismatch.** PageUp/PageDown/Home/End not in code. Either the set is incomplete or the scope was deliberately narrowed.
>
> **Important — `SCROLL_KEYS` is typed `Set<string>`.** The natural type is `ReadonlySet<KeyCode>` where `KeyCode` is the literal union of the actual members.
>
> **Suggestion — `isModifierSeek` duplicates membership logic.** Three sites hard-code the seek-key pair. A second small constant `SEEK_KEYS = new Set(["ArrowLeft", "ArrowRight"])` would dedupe.
>
> **Suggestion — `Set<string>` is overkill for five items.** Array + `.includes` + `as const` produces a literal-union type for free.
>
> **Suggestion — `dispatchKeydown` helper has a too-wide return type.**
>
> **Suggestion — `it.each` tuples are `string[][]` not `[KeyCode][]`.**
>
> ### Recommended improvements
>
> ```ts
> const SCROLL_KEYS = ["Space", "ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight"] as const;
> type ScrollKey = typeof SCROLL_KEYS[number];
> const SCROLL_KEY_SET: ReadonlySet<string> = new Set(SCROLL_KEYS);
> ```
>
> A literal union usable in the `case` labels closes the drift hole; `ReadonlySet` removes the mutation surface.

### comment-analyzer

> ## Summary
>
> Cycle 1 caught most drift but two pieces of residual drift remain — one factually wrong claim about historical behaviour, and one misleading reference to modifier keys that don't apply to the test.
>
> ## Critical Issues
>
> **[A] `c:/Users/Andrew/Spem/src/test/keyboard.test.ts:259-260` — incorrect claim about past behaviour**
>
> ```
> // Plain `/` (no Shift) is not a help-modal trigger; previously it was
> // silently swallowed which broke Firefox quick-find. Must pass through.
> ```
>
> Checked against `origin/main:index.ts` — plain `Slash` was never `preventDefault`'d. The "previously it was silently swallowed" claim is false.
>
> Suggestion: drop the historical sentence entirely. Replace with present-tense rationale only.
>
> **[B] `c:/Users/Andrew/Spem/src/test/keyboard.test.ts:207-209` — irrelevant Alt reference**
>
> ```
> // Cmd/Ctrl+Arrow is the "seek-by-section" handler — must preventDefault
> // so the OS shortcut (macOS jump-word on Cmd, Windows browser-back on
> // Alt; on the iPad path, Cmd is the modifier users press) doesn't fire.
> ```
>
> The bracketed test cases below cover `ctrlKey` and `metaKey` only — there is no `altKey` test. Worse, Alt is a meaningful modifier in this codebase (fine-grain bar seek). Dragging Alt into this comment misleads.
>
> ## Important
>
> **[C] `c:/Users/Andrew/Spem/index.ts:229-235` — "plain (no Cmd/Ctrl)" is ambiguous re: Alt**
>
> The word "plain" naturally reads as "no modifier at all", but the code treats Alt-held keys as "plain". Alt+ArrowLeft will be `preventDefault`'d.
>
> **[D] `c:/Users/Andrew/Spem/src/test/keyboard.test.ts:247-248` — Enter test description mis-attributes the contract**
>
> The assertion is correct (Enter is not `preventDefault`'d), but the rationale obscures a real side effect: pressing Enter on document with focus on a `<button>` ALSO triggers `controls.pause()/play()`.
>
> ## Suggestions
>
> **[E] `c:/Users/Andrew/Spem/index.ts:254-257` — Escape sentence documents pre-existing behaviour.**
>
> **[F] `c:/Users/Andrew/Spem/src/test/keyboard.test.ts:267-269` — Escape rationale doesn't fit the scenario.**
>
> **[G] `c:/Users/Andrew/Spem/src/test/keyboard.test.ts:173,191,199` — ticket reference in test name and headers.** Borderline.
>
> **[H] `c:/Users/Andrew/Spem/src/test/keyboard.test.ts:174-177` — "See #10 for the underlying bug" is a backwards reference.** Per the rule that comments referencing the past belong in the PR description, drop the last sentence.
>
> ## Positive findings
>
> - Present-tense rationale in `index.ts:271-275` is a good model.
> - The IME comment at `keyboard.test.ts:299-301` calls out the failure mode a future refactor would silently introduce.
> - The `SCROLL_KEYS` constant with attached comment block is well-placed.
>
> ## Was all drift caught from cycle 1?
>
> Partially. Two pieces remain: [A] factual error about historical behaviour; [B] off-topic modifier reference. Plus the softer "looking backward" breadcrumb at [H].

**Comment by wainwright1000** [comment]
# VERA-10 Final Synthesis (cycle 2)

Mode: work (Vera ran during initial development, before PR open)
Cycle: 2
Generated: 2026-05-28 13:42
Last run:  2026-05-28 14:00

See also: [Original Report (cycle 2)](https://github.com/wainwmr/spem-player/issues/10#issuecomment-4564050528)
(Earlier cycles: [Synthesis (cycle 1)](https://github.com/wainwmr/spem-player/issues/10#issuecomment-4561299282))

## Summary

Cycle 2 ran on the squashed branch produced by `rebase-parked`. Pass 1 surfaced 8 critical/important findings across the five agents; Bob's triage reduced these to 3 real comment-drift items (10-25/26/27, all in `src/test/keyboard.test.ts`) plus 1 commit-message-vs-code discrepancy (10-24 — PageUp/Home/End named in the squash commit message but absent from the code's `SCROLL_KEYS`). The three comment fixes shipped in commit `5b2a6fc`. 10-24 was reclassified at close-out from "amend the commit message" to "name the actual scope in the PR description"; the squash commit message stays as-is, the durable record of the discrepancy is this synthesis comment plus the PR description Mark sees. Pass 2 ran clean apart from one low-severity wording finding (10-32 — "jump-word" claim in the rewritten Cmd+Arrow comment is technically wrong for body-focus, where Cmd+Arrow is history navigation; deferred, pre-existing in both files). The pass-2 finding is exactly the self-inflicted comment-drift pattern cycle 1 also exhibited — pass-2 catches pass-1's edits when they inherit a pre-existing inaccuracy. No re-run after the close-out edits.

## Findings

### 10-24 — [critical] Commit message overstates SCROLL_KEYS scope (PageUp/PageDown/Home/End)

> **code-reviewer / pr-test-analyzer / silent-failure-hunter / type-design-analyzer, index.ts:236-242 and commit d169f16:**
> The squash commit message lists `Space, ArrowUp, ArrowDown, PageUp, PageDown, Home, End` as the swallowed set. Production code only contains `Space, ArrowUp, ArrowDown, ArrowLeft, ArrowRight`. PageUp/PageDown/Home/End are absent. The contract in the commit message is wider than the code's behaviour.

**Bob's triage:** Real artefact, but the *defect* is in the squash commit message (written today during rebase-parked), not in the code. Cycle 1's scope decision was "Space + 4 arrows" — that decision stands; the iPad bug report names "space or arrows" and the app does not bind PageUp/Home/End. Reclassified at close-out: rather than amending d169f16 (which requires rewriting pushed history with nested rebases the workflow discourages), the PR description will name the actual scope ("Space + 4 arrows"). The Vera synthesis comment on the ticket records the commit-message overstatement durably; reviewers see the PR description, not the individual commit messages, so the practical signal Mark gets is accurate.

**Resolution:** deferred — PR description at publish-merge will name the actual scope.

### 10-25 — [important] Test comment makes false historical claim about plain Slash

> **code-reviewer / comment-analyzer A, src/test/keyboard.test.ts:259-260:**
> "Plain `/` (no Shift) is not a help-modal trigger; previously it was silently swallowed which broke Firefox quick-find." False against `origin/main`: the old `case "Slash"` block only acted when `e.shiftKey` was true and never called `preventDefault()` for plain `/`.

**Bob's triage:** Real comment-to-code drift. Comment lies about history; user-skill rule says past references belong in the PR description anyway. Address now: drop the historical sentence, keep the present-tense rationale (Firefox quick-find).

**Resolution:** addressed (commit 5b2a6fc).

### 10-26 — [important] Test comment drags Alt into Cmd/Ctrl+Arrow rationale

> **comment-analyzer B, src/test/keyboard.test.ts:207-209:**
> "Cmd/Ctrl+Arrow is the 'seek-by-section' handler — must preventDefault so the OS shortcut (macOS jump-word on Cmd, Windows browser-back on Alt; on the iPad path, Cmd is the modifier users press) doesn't fire." The bracketed test cases only cover ctrlKey/metaKey; Alt is a meaningful modifier in this codebase (fine-grain bar seek) and is not what the test is pinning.

**Bob's triage:** Real comment drift. Misleads on both what the test asserts and what the production code does. Address now: rewrite the comment to name only Cmd/Ctrl.

**Resolution:** addressed (commit 5b2a6fc).

### 10-27 — [important] Test top-of-file comment references the ticket as "the underlying bug"

> **comment-analyzer H, src/test/keyboard.test.ts:174-177:**
> "Behavioural assertion via `defaultPrevented`, not by spying on `preventDefault()`... See #10 for the underlying bug." The first sentence is excellent guidance; the "See #10" sentence is a backward breadcrumb that belongs in the PR description, not in code.

**Bob's triage:** Real drift per user-skill rule "comments referencing the past belong in the PR description, not the code". Address now: drop the trailing sentence; keep the rationale.

**Resolution:** addressed (commit 5b2a6fc).

### 10-28 — [important] Non-Element `e.target` early-return loses scroll-prevention

> **silent-failure-hunter I1, index.ts:249-252:**
> `if (!(e.target instanceof Element)) return;` runs before the new preventDefault block. If a keydown fires with `e.target` being `document` or `window` (plausible in iOS Safari with focus on `<body>`), the handler returns before swallowing the scroll key, re-introducing the wiggle.

**Bob's triage:** Pre-existing on `main`; this diff did not introduce or move the guard. Real silent-failure mode in theory, but: (a) we have no repro evidence from a real iPad, (b) moving the SCROLL_KEYS block above the Element guard introduces a different risk (interaction with input-like guard for textareas). Defer to a refactor report entry on `index.ts` for the architectural concern. If field repros surface, file a ticket.

**Resolution:** deferred to refactor report (`wiki/refactor-index.ts.md` if exists, else create at next bug-hunt). Not blocking this PR.

### 10-29 — [important] Test asserts defaultPrevented but not paired action

> **pr-test-analyzer I2, src/test/keyboard.test.ts:192-227:**
> A future refactor that removed the inner Space/ArrowLeft/ArrowRight handlers but kept the SCROLL_KEYS swallow would silently break play/pause/seek while every test in this block continued to pass.

**Bob's triage:** Defensive nit — characterises a hypothetical future refactor, not a live defect. The existing test at `keyboard.test.ts:66-81` covers Space-toggles-playing; arrow seeks are tested elsewhere in the file. Defer additional coverage to a follow-up refactor pass on `keyboard.test.ts`; not worth expanding scope here.

**Resolution:** deferred. Will file a Workbench item if the gap proves significant on a future regression.

### 10-30 — [important] No test for Cmd+Space (Spotlight) or plain F (feedback modal)

> **pr-test-analyzer I1, I3:**
> Cmd+Space pass-through and plain-f-opens-feedback are not covered by the new tests.

**Bob's triage:** Defensive coverage additions. Both behaviours are correct in the code path (Cmd+Space hits no branch and returns; plain F hits the `case "KeyF"` switch and shows feedback). The existing positive Space and negative Cmd+S/Cmd+F tests are representative. Defer.

**Resolution:** deferred.

### 10-31 — [suggestion] Type-design improvements (literal union, ReadonlySet, SEEK_KEYS dedupe)

> **type-design-analyzer:**
> `SCROLL_KEYS: ReadonlySet<KeyCode>` with `as const` literal union would close the typo-vs-switch drift hole. `isModifierSeek` triplicates the seek-key pair across three sites; could share a `SEEK_KEYS` constant.

**Bob's triage:** Genuine refactor opportunities, not defects. The current `Set<string>` works correctly; the proposed literal union is structural improvement. File for the refactor report rather than expanding this PR.

**Resolution:** deferred to refactor report for `index.ts`.

### 10-32 — [suggestion] Pass-2 — "jump-word" wording inaccurate for body-focus path

> **comment-analyzer pass 2, src/test/keyboard.test.ts:206-208 and index.ts:271-275:**
> Both comments cite the OS shortcut as "macOS Cmd+Arrow jump-word" / "the OS jump-word shortcut". Cmd+ArrowLeft/Right is jump-word only when focus is inside a text field. On `<body>` focus (the path the handler actually exercises), macOS Safari/Chrome map Cmd+Arrow to history Back/Forward. The named OS behaviour doesn't match the focus context.

**Bob's triage:** Real factual inaccuracy in user-facing prose (well, comments), but pre-existing in both files (the test comment inherited it via cycle-2 commit 5b2a6fc). Low severity per the gate's framework. Two-line fix would address it but triggers pass 3. Defer — opportunistic fix at the next touch of either comment, or pull into a redo-pr if Mark queries the wording.

**Resolution:** deferred.

### Suggestions (not blocking)

- code-reviewer #2: Prettier collapse of `isPlainScrollKey` — `prettier --check` isn't blocking CI for this file; accept.
- silent-failure I3, S1, S2, S3: cancelable footgun, Alt/Shift+Arrow nuance, `e.code` vs `e.key`, Shift+Space — defer.
- pr-test-analyzer S1-S6: dispatch target consistency, `control`-class guard, Escape-on-input, IME keyCode, key-with-handler-not-in-set, instanceof Element negative case — defer to follow-up.
- comment-analyzer C, D, E, F, G: "plain (no Cmd/Ctrl)" ambiguity, Enter button rationale, Escape sentence placement, Escape rationale, ticket reference in test name — accept as-is or defer.

- Claude
